pkg/controller/login/select_password.go

```diff
@@ -16,53 +16,140 @@
 package login
 
 import (
+	"context"
+	"errors"
+	"fmt"
 	"net/http"
 	"time"
+	"unicode"
 
+	"github.com/google/exposure-notifications-verification-server/internal/firebase"
 	"github.com/google/exposure-notifications-verification-server/pkg/controller"
 	"github.com/google/exposure-notifications-verification-server/pkg/controller/flash"
 )
 
 func (c *Controller) HandleShowSelectNewPassword() http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		ctx := r.Context()
+		session := controller.SessionFromContext(ctx)
+		f := flash.New(session.Values)
 
-		m := controller.TemplateMapFromContext(ctx)
-		m["firebase"] = c.config.Firebase
-		m["requirements"] = &c.config.PasswordRequirements
-		c.h.RenderHTML(w, "login/select-password", m)
+		code := r.FormValue("oobCode")
+		if code == "" {
+			f.Error("No oobCode.")
+			c.renderShowSelectPassword(ctx, w, "", code, f)
+			return
+		}
+
+		email, err := c.firebaseInternal.VerifyPasswordResetCode(ctx, code)
+		if err != nil {
+			if errors.Is(err, firebase.ErrInvalidOOBCode) || errors.Is(err, firebase.ErrExpiredOOBCode) {
+				f.Error("The action code is invalid. This can happen if the code is malformed, expired, or has already been used.")
+			} else {
+				f.Error("Error checking code. %v", err)
+			}
+			c.renderShowSelectPassword(ctx, w, "", code, f)
+			return
+		}
+
+		c.renderShowSelectPassword(ctx, w, email, code, f)
 	})
 }
 
+func (c *Controller) renderShowSelectPassword(ctx context.Context, w http.ResponseWriter, email, code string, f *flash.Flash) {
+	m := controller.TemplateMapFromContext(ctx)
+	m["email"] = email
+	m["code"] = code
+	m["flash"] = f
+	m["requirements"] = &c.config.PasswordRequirements
+	c.h.RenderHTML(w, "login/select-password", m)
+}
+
 func (c *Controller) HandleSubmitNewPassword() http.Handler {
 	logger := c.logger.Named("login.HandleSubmitNewPassword")
 
 	type FormData struct {
-		Email string `form:"email"`
+		Password string `form:"password"`
+		Email    string `form:"email"`
 	}
 
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		ctx := r.Context()
+		session := controller.SessionFromContext(ctx)
+		f := flash.New(session.Values)
```

Comment:
There's a controller helper for this. I think it's FlashFromSession. Also we should name this `flash` to be consistent.